cpp/ql/src/Documentation/CommentedOutCode.qll

```diff
@@ -2,45 +2,57 @@ import cpp
- *  * HTML entities in common notation (e.g. &amp;gt; and &amp;eacute;)
- *  * HTML entities in decimal notation (e.g. a&amp;#768;)
- *  * HTML entities in hexadecimal notation (e.g. &amp;#x705F;)
- * To account for the code generated by protobuf, we also insist that the comment
- * does not begin with `optional` or `repeated` and end with a `;`, which would
- * normally be a quoted bit of literal `.proto` specification above the associated
- * declaration.
- * To account for emacs folding markers, we ignore any line containing
- * `{{{` or `}}}`.
- *
- * Finally, some code tends to embed GUIDs in comments, so we also exclude those.
  */
 bindingset[line]
 private predicate looksLikeCode(string line) {
-    exists(string trimmed |
-           trimmed = line.regexpReplaceAll("(?i)(^\\s+|&#?[a-z0-9]{1,31};|\\s+$)", "") |
-           trimmed.regexpMatch(".*[{};]")
-           and (
-              // If this line looks like code because it ends with a closing
-              // brace that's preceded by something other than whitespace ...
-              trimmed.regexpMatch(".*.\\}")
-              implies
-              // ... then there has to be ") {" (or some variation)
-              // on the line, suggesting it's a statement like `if`
-              // or a function declaration. Otherwise it's likely to be a
-              // benign use of braces such as a JSON example or explanatory
-              // pseudocode.
-              trimmed.regexpMatch(".*(\\)|const|volatile|override|final|noexcept|&)\\s*\\{.*")
-           )
-           and not trimmed.regexpMatch("(>.*|.*[\\\\@][{}].*|(optional|repeated) .*;|.*(\\{\\{\\{|\\}\\}\\}).*|\\{[-0-9a-zA-Z]+\\})"))
+  exists(string trimmed |
+    // trim leading and trailing whitespace, and HTML codes: 
+    //  * HTML entities in common notation (e.g. &amp;gt; and &amp;eacute;)
+    //  * HTML entities in decimal notation (e.g. a&amp;#768;)
+    //  * HTML entities in hexadecimal notation (e.g. &amp;#x705F;)
+    trimmed = line.regexpReplaceAll("(?i)(^\\s+|&#?[a-z0-9]{1,31};|\\s+$)", "")
+  |
+    (
+      (
+        // Match comment lines ending with '{', '}' or ';'
+        trimmed.regexpMatch(".*[{};]") and
+        (
+          // If this line looks like code because it ends with a closing
+          // brace that's preceded by something other than whitespace ...
+          trimmed.regexpMatch(".*.\\}")
+          implies
+          // ... then there has to be ") {" (or some variation)
+          // on the line, suggesting it's a statement like `if`
+          // or a function declaration. Otherwise it's likely to be a
+          // benign use of braces such as a JSON example or explanatory
+          // pseudocode.
+          trimmed.regexpMatch(".*(\\)|const|volatile|override|final|noexcept|&)\\s*\\{.*")
+        )
+      ) or (
+        // Match comment lines that look like preprocessor code
+        trimmed.regexpMatch("#(include|define|undef|if|ifdef|ifndef|elif|else|endif|error)(\\s.*|)")
```

Comment:
^ For the sake of reviewers, this line is the meaningful change.

Reply:
The tokenizer is run before the preprocessor, which can have some weird effects: https://godbolt.org/z/pA4W3s

Reply:
I would also add <code>#pragma</code> to the preprocessor regexp.  Also, it is possible to have whitespace between the <code>#</code> and a directive.

Also, s/<code>function declaration</code>/<code>function definition</code>, since you have an opening brace.  (At least that's the terminology I grew up with.)

Reply:
I think what @rdmarsh2 suggests is to replace `(\\s.*|)` with `\\b.*` (`\b` matches a zero-width word boundary).

Reply:
Thanks for the suggestions - all implemented.  I wasn't aware of `\b` before now, very useful.